### ISSUE TITLE
Strongly normalizing IR

### ISSUE THREAD

**Issue by jamii** [MaterializeInc/materialize#692]
Many of our transforms are an attempt to reach some kind of normal form, with joins/filters/maps/projects etc fused and pushed down as far as they can go. 

We could instead use a logical IR which doesn't specify the order of these operations:

```
enum LogicalRelationExpr {
  Predicate{
    inputs: Vec<LogicalRelationExpr>,
    maps: Vec<ScalarExpr>,
    filters: Vec<ScalarExpr>,
    project: Vec<usize>,
  }
  Reduce{...}
  Union{...}
  Branch{...}
}
```

Then in the transformation to the physical IR, we have one function that greedily takes inputs, maps, filters and projects from the logical IR.

**Comment by benesch** [comment]
I believe we got this with the map-filter-project operator (#6339).

**Comment by frankmcsherry** [comment]
@asenac may want to revive this, as he is specifically interested in a multilinear operator as proposed here. That would be something we do not yet have.

**Comment by asenac** [comment]
Yes! This is basically the idea behind the query graph model. Normalization/simplification is a lot easier/simpler with a reduced set of logical operations. I think we should revive this and do it after the LIR is pushed. Then, we can fuse these operators into one and move join planning to the code that does the MIR -> LIR transformation.
